In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

C:\Users\Yan Peng\anaconda3\Lib\site-packages\pandas\core\arrays\masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (


# 1. Import data

In [3]:
path = r'C:\Users\Yan Peng\LA Crimes'

In [4]:
df = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'Crime_Data_from_2020_to_Present.csv'),  index_col = False)

# 2. Data consistency checks

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 883987 entries, 0 to 883986
Data columns (total 28 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   DR_NO           883987 non-null  int64  
 1   Date Rptd       883987 non-null  object 
 2   DATE OCC        883987 non-null  object 
 3   TIME OCC        883987 non-null  int64  
 4   AREA            883987 non-null  int64  
 5   AREA NAME       883987 non-null  object 
 6   Rpt Dist No     883987 non-null  int64  
 7   Part 1-2        883987 non-null  int64  
 8   Crm Cd          883987 non-null  int64  
 9   Crm Cd Desc     883987 non-null  object 
 10  Mocodes         760974 non-null  object 
 11  Vict Age        883987 non-null  int64  
 12  Vict Sex        766927 non-null  object 
 13  Vict Descent    766919 non-null  object 
 14  Premis Cd       883977 non-null  float64
 15  Premis Desc     883447 non-null  object 
 16  Weapon Used Cd  306604 non-null  float64
 17  Weapon Des

In [6]:
# Change column names to comprehensible ones

new_column_names = {'DR_NO': 'case_number', 'Date Rptd': 'date_reported', 'DATE OCC': 'date_occurred', 'TIME OCC': 'time_occurred',
                    'AREA': 'area', 'AREA NAME': 'area_name', 'Rpt Dist No': 'reporting_distict', 'Part 1-2': 'part_1_2_crimes', 'Crm Cd': 'crime_code', 'Crm Cd Desc': 'crime_code_description',
                    'Vict Age': 'victim_age', 'Vict Sex': 'victim_sex', 'Vict Descent': 'victim_descent', 'Premis Cd': 'premise_code', 'Premis Desc': 'premise_description',
                    'Weapon Used Cd': 'weapon_used_code', 'Weapon Desc': 'weapon_description', 'LOCATION': 'location'}
df.rename(columns=new_column_names, inplace=True)

In [7]:
# Checking for mixed-type data

for col in df.columns.tolist():
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)

C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1890984043.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1890984043.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1890984043.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1890984043.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\18909840

Mocodes


C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1890984043.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1890984043.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)


victim_sex
victim_descent


C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1890984043.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1890984043.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1890984043.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1890984043.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)


premise_description


C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1890984043.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1890984043.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)


weapon_description


C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1890984043.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1890984043.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1890984043.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1890984043.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\18909840

Cross Street


C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1890984043.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)


In [8]:
# Check sex column

df['victim_sex'].value_counts()

victim_sex
M    363260
F    323891
X     79680
H        95
-         1
Name: count, dtype: int64

#### M and F stand for male and female, respectively. The other values will be considered unknown.

In [9]:
# Check descent column

df['victim_descent'].value_counts()

victim_descent
H    269743
W    179133
B    124820
X     87776
O     69907
A     19377
K      4788
F      3765
C      3494
J      1238
V       935
I       840
Z       447
P       237
U       176
D        72
G        64
L        57
S        48
-         2
Name: count, dtype: int64

#### H: Hispanic, W: White, B: Black, X: Unknown, O: Other, A: Asian

In [10]:
# Summary statistics

df.describe()

case_number  time_occurred           area  reporting_distict  \
count  8.839870e+05  883987.000000  883987.000000      883987.000000   
mean   2.173618e+08    1336.560971      10.703108        1116.747620   
std    1.139578e+07     653.232813       6.100755         610.076564   
min    8.170000e+02       1.000000       1.000000         101.000000   
25%    2.103145e+08     900.000000       6.000000         615.000000   
50%    2.203222e+08    1415.000000      11.000000        1141.000000   
75%    2.302216e+08    1900.000000      16.000000        1615.000000   
max    2.499046e+08    2359.000000      21.000000        2199.000000   

       part_1_2_crimes     crime_code     victim_age   premise_code  \
count    883987.000000  883987.000000  883987.000000  883977.000000   
mean          1.411516     500.758133      29.660630     306.299510   
std           0.492109     207.629796      21.825831     217.174128   
min           1.000000     110.000000      -3.000000     101.000000   
25%           1.000000     331.000000       0.000000     101.000000   
50%           1.000000     442.000000      31.000000     203.000000   
75%           2.000000     626.000000      45.000000     501.000000   
max           2.000000     956.000000     120.000000     976.000000   

       weapon_used_code       Crm Cd 1      Crm Cd 2     Crm Cd 3    Crm Cd 4  \
count     306604.000000  883976.000000  64497.000000  2176.000000   62.000000   
mean         363.322520     500.499365    957.820379   983.770221  990.983871   
std          123.705678     207.419239    110.885829    52.538656   27.477259   
min          101.000000     110.000000    210.000000   310.000000  821.000000   
25%          310.000000     331.000000    998.000000   998.000000  998.000000   
50%          400.000000     442.000000    998.000000   998.000000  998.000000   
75%          400.000000     626.000000    998.000000   998.000000  998.000000   
max          516.000000     956.000000    999.000000   999.000000  999.000000   

                 LAT            LON  
count  883987.000000  883987.000000  
mean       33.986470    -118.051153  
std         1.725364       5.981531  
min         0.000000    -118.667600  
25%        34.014300    -118.429700  
50%        34.058500    -118.321500  
75%        34.163500    -118.273900  
max        34.334300       0.000000

#### There are no unusual values except for the minus values in the category of age and the max. value of 120.

In [11]:
df['victim_age'].value_counts().sort_index()

victim_age
-3           2
-2          14
-1          67
 0      221363
 2         399
         ...  
 96         91
 97         65
 98         68
 99        313
 120         1
Name: count, Length: 103, dtype: int64

#### These extreme values make up less than 100 rows and can thus be ignored.

In [12]:
# Finding missing values

df.isnull().sum()

case_number                    0
date_reported                  0
date_occurred                  0
time_occurred                  0
area                           0
area_name                      0
reporting_distict              0
part_1_2_crimes                0
crime_code                     0
crime_code_description         0
Mocodes                   123013
victim_age                     0
victim_sex                117060
victim_descent            117068
premise_code                  10
premise_description          540
weapon_used_code          577383
weapon_description        577383
Status                         0
Status Desc                    0
Crm Cd 1                      11
Crm Cd 2                  819490
Crm Cd 3                  881811
Crm Cd 4                  883925
location                       0
Cross Street              744208
LAT                            0
LON                            0
dtype: int64

#### The sex of about one eighth of the victims is unknown. Otherwise there are no significant missing values relevant for the analysis.

In [13]:
# Finding duplicates

df_dups = df[df.duplicated()]

In [14]:
df_dups

Empty DataFrame
Columns: [case_number, date_reported, date_occurred, time_occurred, area, area_name, reporting_distict, part_1_2_crimes, crime_code, crime_code_description, Mocodes, victim_age, victim_sex, victim_descent, premise_code, premise_description, weapon_used_code, weapon_description, Status, Status Desc, Crm Cd 1, Crm Cd 2, Crm Cd 3, Crm Cd 4, location, Cross Street, LAT, LON]
Index: []

[0 rows x 28 columns]

#### There are no duplicates.

# 3. Export data

In [15]:
# Write out abbreviations and set all letters in uppercase to align with the names in the geojson file

df['area_name'].replace('N Hollywood', 'North Hollywood', inplace=True)
df['area_name'].replace('West LA', 'West Los Angeles', inplace=True)
df['area_name'] = df['area_name'].str.upper()

C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1662919639.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['area_name'].replace('N Hollywood', 'North Hollywood', inplace=True)
C:\Users\Yan Peng\AppData\Local\Temp\ipykernel_2224\1662919639.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always 

In [16]:
df['area_name']

0            WILSHIRE
1             CENTRAL
2           SOUTHWEST
3            VAN NUYS
4           HOLLYWOOD
             ...     
883982    77TH STREET
883983      HOLLYWOOD
883984        CENTRAL
883985       VAN NUYS
883986      SOUTHWEST
Name: area_name, Length: 883987, dtype: object

In [18]:
df.head()

case_number           date_reported           date_occurred  time_occurred  \
0    190326475  03/01/2020 12:00:00 AM  03/01/2020 12:00:00 AM           2130   
1    200106753  02/09/2020 12:00:00 AM  02/08/2020 12:00:00 AM           1800   
2    200320258  11/11/2020 12:00:00 AM  11/04/2020 12:00:00 AM           1700   
3    200907217  05/10/2023 12:00:00 AM  03/10/2020 12:00:00 AM           2037   
4    220614831  08/18/2022 12:00:00 AM  08/17/2020 12:00:00 AM           1200   

   area  area_name  reporting_distict  part_1_2_crimes  crime_code  \
0     7   WILSHIRE                784                1         510   
1     1    CENTRAL                182                1         330   
2     3  SOUTHWEST                356                1         480   
3     9   VAN NUYS                964                1         343   
4     6  HOLLYWOOD                666                2         354   

                     crime_code_description  ... Status   Status Desc  \
0                          VEHICLE - STOLEN  ...     AA  Adult Arrest   
1                     BURGLARY FROM VEHICLE  ...     IC   Invest Cont   
2                             BIKE - STOLEN  ...     IC   Invest Cont   
3  SHOPLIFTING-GRAND THEFT ($950.01 & OVER)  ...     IC   Invest Cont   
4                         THEFT OF IDENTITY  ...     IC   Invest Cont   

  Crm Cd 1 Crm Cd 2  Crm Cd 3 Crm Cd 4  \
0    510.0    998.0       NaN      NaN   
1    330.0    998.0       NaN      NaN   
2    480.0      NaN       NaN      NaN   
3    343.0      NaN       NaN      NaN   
4    354.0      NaN       NaN      NaN   

                                   location Cross Street      LAT       LON  
0   1900 S  LONGWOOD                     AV          NaN  34.0375 -118.3506  
1   1000 S  FLOWER                       ST          NaN  34.0444 -118.2628  
2   1400 W  37TH                         ST          NaN  34.0210 -118.3002  
3  14000    RIVERSIDE                    DR          NaN  34.1576 -118.4387  
4                         1900    TRANSIENT          NaN  34.0944 -118.3277  

[5 rows x 28 columns]

In [17]:
df.to_csv(os.path.join(path, '02 Data','Prepared Data', 'crime_data_checked.csv'))